### ISSUE TITLE
TreeGrid Drag and Drop: Client side drop validation seems unfinished

### ISSUE THREAD

**Issue by kaushikvijay** [vaadin/framework#9834]
- Vaadin version:8.1.2
- Browser: Chome 60.0.3112.78 (working in FF:52.2)
- Description of the bug: 
I am trying to restrict drop of the a leaf node on some nodes while doing drag and drop at client side with TreeGrid. I need to let user know in advance(on hover/dragover) that certain drop is possible or not.(In vaadin 7, we were using  `ClientSideCriterion ` to achieve same). So I need to know at client end that the target node is leaf or not. So I set this data on DragStart using `DragDataGenerator ` or `GridDragStartListener  ` as below
```
        TreeGrid<TOC> tocTree = new TreeGrid<>()
       // some config for tree ....
        TreeGridDragSource<TOC> dragSource = new TreeGridDragSource<>(tocTree);
        dragSource.setEffectAllowed(EffectAllowed.MOVE);
        dragSource.setDragDataGenerator("nodetype", tocVO -> {//for client side
            String dragType = tocVO.getType().getName();
            return dragType;
        });
```

Now, i want to check this data on client end to show indication that drop is allowed/ disallowed. I am trying to use `setDropCriteriaScript` for this but I am unable to get data due to limitations by Chrome.
```
       TreeGridDropTarget<TOC> dropTarget = new TreeGridDropTarget<>(tocTree, DropMode.ON_TOP_OR_BETWEEN);
        dropTarget.setDropEffect(DropEffect.MOVE);
        dropTarget.setDropCriteriaScript(
                        + "if ( event.dataTransfer.types.includes && event.dataTransfer.types.includes('nodetype')) {"
                        +       "var nodetype= event.dataTransfer.getData('nodetype');"//NOT GETTING DATA
                        +       "if(nodetype!=='leaf'){"
                        +          "return true;"
                        +       "}"
                        + "}"
                        // Otherwise cancel the event
                        + "return false;"
                    );
```
But on `dropover `events,  `dataTransfer `is protected mode which doesn't allow reading of data. So this API is useless to do any meaningful client side validation. We need to know `dropTargetType `and `dragSourceType `to do any meaningful validation. Using Criterian class is also not producing any meaningful results as Criterian is applicable to whole tree but not to a particular node.
So it feels drag and drop functionality is not complete for treegrid and a better API to do ClientSide validation is required.

**Comment by kauvj** [comment]
A comment on roadmap would be helpful so that we can decide to go forward with different approach.

**Comment by deformedpepper** [comment]
What does the different approach look like? We need to implement per row validation for the Grid too.

**Comment by kauvj** [comment]
Unfortunately, we are living with server side validation which is not a good solution for our use case. Next step would be using something in pure JavaScript. Implementing something in GWT din't bring  benefit in past as with every vaadin upgrade, our custom components became useless and we had to spend considerable time resolving already solved issues.

**Comment by TatuLund** [comment]
Should be fixed in Vaadin 8.10.3 by https://github.com/vaadin/framework/pull/11931
